packages/protocol/contracts/L1/libs/LibProving.sol

```diff
@@ -6,372 +6,199 @@
-        TaikoData.Config memory config,
-        AddressResolver resolver,
-        IProofVerifier proofVerifier,
-        TaikoData.Evidence memory evidence,
-        TaikoData.BlockMetadata memory target,
-        bytes32 blockHashOverride
-    ) private {
-        if (evidence.meta.id != target.id) revert L1_ID();
-        if (evidence.prover == address(0)) revert L1_PROVER();
+        TaikoData.BlockEvidence memory evidence
+    ) internal {
+        TaikoData.BlockMetadata memory meta = evidence.meta;
+        if (
+            meta.id != blockId ||
+            meta.id <= state.lastBlockId ||
+            meta.id >= state.nextBlockId
+        ) revert L1_ID();
 
-        if (FLAG_CHECK_METADATA) {
-            if (
-                target.id <= state.latestVerifiedId ||
-                target.id >= state.nextBlockId
-            ) revert L1_ID();
-            if (
-                state
-                    .getProposedBlock(config.maxNumBlocks, target.id)
-                    .metaHash != target.hashMetadata()
-            ) revert L1_META_MISMATCH();
-        }
+        if (
+            // 0 and 1 (placeholder) are not allowed
+            uint256(evidence.parentHash) <= 1 ||
+            // 0 and 1 (placeholder) are not allowed
+            uint256(evidence.blockHash) <= 1 ||
+            // cannot be the same hash
+            evidence.blockHash == evidence.parentHash ||
+            // 0 and 1 (placeholder) are not allowed
+            uint256(evidence.signalRoot) <= 1 ||
+            // prover must not be zero
+            evidence.prover == address(0)
+        ) revert L1_INVALID_EVIDENCE();
+
+        TaikoData.ProposedBlock storage proposal = state.proposedBlocks[
+            meta.id % config.maxNumBlocks
+        ];
 
-        if (FLAG_VALIDATE_HEADER_FOR_METADATA) {
-            if (
-                evidence.header.parentHash == 0 ||
-                evidence.header.beneficiary != evidence.meta.beneficiary ||
-                evidence.header.difficulty != 0 ||
-                evidence.header.gasLimit !=
-                evidence.meta.gasLimit + config.anchorTxGasLimit ||
-                evidence.header.gasUsed == 0 ||
-                evidence.header.timestamp != evidence.meta.timestamp ||
-                evidence.header.extraData.length !=
-                evidence.meta.extraData.length ||
-                keccak256(evidence.header.extraData) !=
-                keccak256(evidence.meta.extraData) ||
-                evidence.header.mixHash != evidence.meta.mixHash
-            ) revert L1_META_MISMATCH();
-        }
+        if (proposal.metaHash != LibUtils.hashMetadata(meta))
+            revert L1_EVIDENCE_MISMATCH();
 
-        // For alpha-2 testnet, the network allows any address to submit ZKP,
-        // but a special prover can skip ZKP verification if the ZKP is empty.
+        TaikoData.ForkChoice storage fc = state.forkChoices[
+            blockId % config.maxNumBlocks
+        ][proposal.nextForkChoiceId];
```

Comment:
Based on these lines (`fc` instance creation) + the check
`if (uint256(fc.chainData.blockHash) <= 1)`
it seems to me we are not restricting proofs to be submitted multiple times, because `nextForkChoiceId` is always incremented (after success), so that the if above will always check against a new (empty) instance of `fc` ?

Reply:
> I may be misunderstanding this currently, but this looks a bit strange. We get the data at `nextForkChoiceId`, but I don't understand why we use this to check if it's already been proven. Shouldn't it use `state.forkChoiceIds[blockId][evidence.parentHash]` to do that (so we accept only the first proof for each "fork")? The other strange thing is that `state.forkChoiceIds[blockId][evidence.parentHash]` is overwritten at the end of this code block without any other place using `evidence.parentHash`, so it seems like this value could simply be overwritten by a proof having the same `parentHash` but that is submitted later that passes through all the checks on `state.forkChoices[blockId % config.maxNumBlocks][proposal.nextForkChoiceId]`. I see only the correct fork choice being reset after the block is verified, so it seems like these `forkChoices` could contain old data that would still be used when the block slot is reused.

I think you've found an implementation bug, I try to fix it here: https://github.com/taikoxyz/taiko-mono/pull/13387